libsolidity/experimental/analysis/TypeClassMemberRegistration.cpp

```diff
@@ -0,0 +1,139 @@
+
+#include <libsolidity/experimental/analysis/Analysis.h>
+#include <libsolidity/experimental/analysis/TypeClassRegistration.h>
+#include <libsolidity/experimental/analysis/TypeRegistration.h>
+#include <libsolidity/experimental/ast/TypeSystemHelper.h>
+
+#include <liblangutil/ErrorReporter.h>
+#include <liblangutil/Exceptions.h>
+
+using namespace solidity::frontend::experimental;
+using namespace solidity::langutil;
+
+TypeClassMemberRegistration::TypeClassMemberRegistration(Analysis& _analysis):
+	m_analysis(_analysis),
+	m_errorReporter(_analysis.errorReporter()),
+	m_typeSystem(_analysis.typeSystem())
+{
+	TypeSystemHelpers helper{m_typeSystem};
+
+	auto registeredTypeClass = [&](BuiltinClass _builtinClass) -> TypeClass {
+		return m_analysis.annotation<TypeClassRegistration>().builtinClasses.at(_builtinClass);
+	};
+
+	auto defineConversion = [&](BuiltinClass _builtinClass, PrimitiveType _fromType, std::string _functionName) {
+		annotation().typeClassFunctions[registeredTypeClass(_builtinClass)] = {{
+			std::move(_functionName),
+			helper.functionType(
+				m_typeSystem.type(_fromType, {}),
+				m_typeSystem.typeClassInfo(registeredTypeClass(_builtinClass)).typeVariable
+			),
+		}};
+	};
+
+	auto defineBinaryMonoidalOperator = [&](BuiltinClass _builtinClass, Token _token, std::string _functionName) {
+		Type typeVar = m_typeSystem.typeClassInfo(registeredTypeClass(_builtinClass)).typeVariable;
+		annotation().operators.emplace(_token, std::make_tuple(registeredTypeClass(_builtinClass), _functionName));
+		annotation().typeClassFunctions[registeredTypeClass(_builtinClass)] = {{
+			std::move(_functionName),
+			helper.functionType(
+				helper.tupleType({typeVar, typeVar}),
+				typeVar
+			)
+		}};
+	};
+
+	auto defineBinaryCompareOperator = [&](BuiltinClass _builtinClass, Token _token, std::string _functionName) {
+		Type typeVar = m_typeSystem.typeClassInfo(registeredTypeClass(_builtinClass)).typeVariable;
+		annotation().operators.emplace(_token, std::make_tuple(registeredTypeClass(_builtinClass), _functionName));
+		annotation().typeClassFunctions[registeredTypeClass(_builtinClass)] = {{
+			std::move(_functionName),
+			helper.functionType(
+				helper.tupleType({typeVar, typeVar}),
+				m_typeSystem.type(PrimitiveType::Bool, {})
+			)
+		}};
+	};
+
+	defineConversion(BuiltinClass::Integer, PrimitiveType::Integer, "fromInteger");
+
+	defineBinaryMonoidalOperator(BuiltinClass::Mul, Token::Mul, "mul");
+	defineBinaryMonoidalOperator(BuiltinClass::Add, Token::Add, "add");
+
+	defineBinaryCompareOperator(BuiltinClass::Equal, Token::Equal, "eq");
+	defineBinaryCompareOperator(BuiltinClass::Less, Token::LessThan, "lt");
+	defineBinaryCompareOperator(BuiltinClass::LessOrEqual, Token::LessThanOrEqual, "leq");
+	defineBinaryCompareOperator(BuiltinClass::Greater, Token::GreaterThan, "gt");
+	defineBinaryCompareOperator(BuiltinClass::GreaterOrEqual, Token::GreaterThanOrEqual, "geq");
```

Comment:
@matheusaaguiar The task for you now, on top of this PR, would be to remove the whole concept of `BuiltinClass` and instead make it possible to register these members using `__builtin()`. This is the remaining part of pass 3 from Daniel's note.

To do this, you'd remove this hard-coded initialization and only fill in the `typeClassFunctions`/`operators` annotations when visiting `Builtin` AST node. Of course only if the usage is correct - only one definition, only those predefined operators/conversions, etc. After this change we'll also no longer need the ability to use a token as type class name, so I'd remove that too (otherwise it will just be dead code).

You'll also need a small parser/AST modification to allow more than one argument in `__builtin()` (in #14566 I made it expect exactly one).

When you do this, the code that currently looks like this:
```solidity
pragma experimental solidity;

type word = __builtin("word");

type uint256 = word;

instantiation uint256: + {
    function add(x: uint256, y: uint256) -> uint256 {}
}

function f(x: uint256, y: uint256) -> uint256 {
    return x + y;
}
```

should look like this:
```solidity
pragma experimental solidity;

type word = __builtin("word");

type uint256 = word;

__builtin("+", Add.add);

class uint256: Add {
    function add(x: uint256, y: uint256) -> uint256;
}

instantiation uint256: Add {
    function add(x: uint256, y: uint256) -> uint256 {}
}

function f(x: uint256, y: uint256) -> uint256 {
    return x + y;
}
```

Without `__builtin()` most of this should still work, but the `x + y` bit should report an error due to the operator `+` not being defined.